### PR TITLE
GPU fusion code cleanup: Extract GpuInstructionFusion::IsFusible into gpu_fusible.cc

### PR DESCRIPTION
**Pull request by sana-damani**

### CODE REVIEW

**Comment by sana-damani** [comment]
I have signed the CLA

**Review by thomasjoerg — DISMISSED** [review]

**Comment by rthadur** [comment]
@sana-damani please sign CLA

**Comment by sana-damani** [comment]
https://cla.developers.google.com/clas shows that I signed it on Feb 18.

Sana

**Comment by sanjoy** [comment]
@rthadur Any idea what we're waiting on here?

**Comment by sana-damani** [comment]
> @rthadur Any idea what we're waiting on here?

Looks like it needs another review before it can be merged. This is likely because I committed a new change before the merge was done.

**Review by sanjoy — DISMISSED** [review]

**Comment by sanjoy** [comment]
@rthadur I still see the "Need a CLA for one or more commit authors" line even though @sana-damani says they've signed the CLA.

**Comment by rthadur** [comment]
> @rthadur I still see the "Need a CLA for one or more commit authors" line even though @sana-damani says they've signed the CLA.

@sanjoy  i believe there were multiple people who contributed to this PR , @sana-damani do you know if there is one more contributor for this PR ?

**Comment by sana-damani** [comment]
I was the only contributor.

**Comment by yifeif** [comment]
Hi @sana-damani, did you use the same username and email address for your commits? I see "sdamani" as the username on the commits.

**Comment by sana-damani** [comment]
Thanks for pointing that out! I'll try and fix the commit author.

**Comment by sana-damani** [comment]
I have signed the CLA with the updated username

**Review by sanjoy — DISMISSED** [review]

**Comment by sanjoy** [comment]
I can see that sdamani with email s*******@gatech.edu has signed the CLA (but not sana-damani); maybe the system is not doing the right thing because your github username is sana-damani and sdamani?

**Comment by thomasjoerg** [comment]
@sana-damani The last message from googlebot was sent because different emails appear in the commit history (s...@gatech.edu and s...@gmail.com). I can see that you signed the CLA with each of them, so I'm not entirely sure what's going on. In any case, it might be easier to stick to a single email address going forward. You're PR is being reviewed internally.

**Comment by sana-damani** [comment]
Is the Windows Bazel GPU build failure a real failure? If so, is there a way that I can reproduce and debug this failure?

**Comment by sanjoy** [comment]
I don't think the windows failure is related.

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 1]
```diff
@@ -131,5 +131,41 @@ bool ShapesCompatibleForMultiOutputFusion(const HloInstruction& instr1,
                                              get_loop_shape(instr_2));
 }
 
+bool IsInputFusible(const HloInstruction& instr) {
+  // Input fusion only handles non-elemental reduction and scatter operations.
+  return IsInputFusibleReduction(instr) ||
+         instr.opcode() == HloOpcode::kScatter;
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: This breaks InstructionFusionTest.FuseIntoScatter. The function works for unfused Scatter ops but not for existing Scatter input fusions. You want to add something like this.

         (instr.opcode() == HloOpcode::kFusion &&
          instr.fusion_kind() == HloInstruction::FusionKind::kInput &&
          instr.fused_expression_root()->opcode() == HloOpcode::kScatter)

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.cc [hunk 2]
```diff
@@ -150,8 +150,7 @@ bool IsLoopFusible(const HloInstruction& instr) {
          instr.opcode() == HloOpcode::kConcatenate ||
          instr.opcode() == HloOpcode::kDynamicSlice ||
          instr.opcode() == HloOpcode::kDynamicUpdateSlice ||
-         (instr.opcode() == HloOpcode::kFusion &&
-          instr.fusion_kind() == HloInstruction::FusionKind::kLoop) ||
+         instr.opcode() == HloOpcode::kFusion ||
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Existing input fusions are not loop-fusible, hence we should return false as we did before. Please update IsInputFusible as suggested in my other comment instead.

Comment (sana-damani) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: What about output and custom fusion nodes? Should they also be disallowed from loop fusion?

Comment (thomasjoerg) [reply tensorflow/compiler/xla/service/gpu/gpu_fusible.cc]: Yes, that's a good point. We want to check for FusionKind::kLoop here.

File: tensorflow/compiler/xla/service/gpu/gpu_fusible.h [hunk 3]
```diff
@@ -46,6 +55,10 @@ bool IsReduceInputFusion(const HloInstruction& instr);
 // is either an unfused reduction-to-vector op or a reduce input fusion.
 bool IsInputFusibleReduction(const HloInstruction& instr);
 
+// Whether `instr` is fusible as root of a reduce input fusions, i.e. `instr`
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/gpu_fusible.h]: Please replace "reduce" with "scatter".

**Review by thomasjoerg — DISMISSED** [review]

**Review by thomasjoerg — APPROVED** [review]

**Pull request merged**